Translate tidb computing (#3057)
pingcap/docs: tidb-computing.md

## SQL layer overview

TiDB's SQL layer, TiDB Server, translates SQL statements into Key-Value operations, forwards the operations to TiKV, the distributed Key-Value storage layer, assembles the results returned by TiKV, and finally returns the query results to the client.

The nodes at this layer are stateless. These nodes themselves do not store data and are completely equivalent.

### SQL computing

The simplest solution to SQL computing is the [mapping of table data to Key-Value](#mapping-of-table-data-to-key-value) as described in the previous section, which maps SQL queries to KV queries, acquires the corresponding data through the KV interface, and performs various computations.

For example, to execute the `select count(*) from user where name = "TiDB"` SQL statement, TiDB needs to read all data in the table, then checks whether the `name` field is `TiDB`, and if so, returns this row. The process is as follows:

> 1. Construct the Key Range: all `RowID` in a table are in `[0,  MaxInt64)` range. According to the row data `Key` encoding rule, using `0` and `MaxInt64` can construct a `[StartKey, EndKey)` range that is left-inclusive and right-exclusive.

Reviewer: ```suggestion
1. Construct the Key Range: all `RowID` in a table are in `[0,  MaxInt64)` range. According to the row data `Key` encoding rule, using `0` and `MaxInt64` can construct a `[StartKey, EndKey)` range that is left-closed and right-open.
```

See https://en.wikipedia.org/wiki/Interval_(mathematics)#Terminology

---

> title: TiDB database computation

Reviewer 2: ```suggestion
title: Computation of TiDB Database
```

summary: Understand the computation layer of the TiDB database.

> category: introduction

Reviewer 2: ```suggestion
category: introduction
aliases: ['/docs/dev/tidb-computing/']
```

---

> # Computation of TiDB database

Reviewer 2: ```suggestion
# Computation of TiDB Database
```

### Example of Key-Value mapping relationship

Finally, a simple example is used to understand the Key-Value mapping relationship of TiDB. Suppose the following table exists in TiDB.

```sql
CREATE TABLE User {
     ID int,
     Name varchar(20),
     Role varchar(20),
     Age int,
     PRIMARY KEY (ID),
     KEY idxAge (Age)
};
```

Suppose there are 3 rows of data in the table.

```
1, "TiDB", "SQL Layer", 10
2, "TiKV", "KV Engine", 20
3, "PD", "Manager", 30
```

> First, each row of data is mapped to a (Key, Value) key-value pair, and the table has an `int` type primary key, so the value of `RowID` is the value of this primary key. Suppose the table has `TableID` of 10, then its table data stored on TiKV is:

Reviewer 2: ```suggestion
Each row of data is mapped to a (Key, Value) key-value pair, and the table has an `int` type primary key, so the value of `RowID` is the value of this primary key. Suppose the table's `TableID` is 10, and then its table data stored on TiKV is:
```

```
t10_r1 --> ["TiDB", "SQL  Layer", 10]
t10_r2 --> ["TiKV", "KV  Engine", 20]
t10_r3 --> ["PD", " Manager", 30]
```

> In addition to the primary key, the table has a non-unique ordinary secondary index, `idxAge`. Suppose the `IndexID` is 1, then its index data stored on TiKV is:

Reviewer 2: ```suggestion
In addition to the primary key, the table has a non-unique ordinary secondary index, `idxAge`. Suppose the `IndexID` is `1`, and then its index data stored on TiKV is:
```

```
t10_i1_10_1 --> null
t10_i1_20_2 --> null
t10_i1_30_3 --> null
```

> The above example shows the mapping rule from a relational model to a Key-Value model in TiDB, and the consideration behind the selection of this scheme.

Reviewer 2: ```suggestion
The above example shows the mapping rule from a relational model to a Key-Value model in TiDB, and the consideration behind this mapping scheme.
```

> ## Meta-information management

Reviewer 2: `元信息` can be translated as `metadata`.  See [TiDB Internal (II) - Computing](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map).
```suggestion
## Metadata management
```

> Each `Database` and `Table` in TiDB has meta information, aka its definition and various attributes. This information also needs to be persistent, and TiDB stores this information in TiKV as well.

Reviewer 2: It is not recommended to use Latin abbreviations in technical documents according to Google technical documentation style guide.
```suggestion
Each database and table in TiDB has metadata that indicates its definition and various attributes. This information also needs to be persisted, and TiDB stores this information in TiKV as well.
```

> Each `Database` / `Table` is assigned an unique ID. As the unique identifier, when encoded as Key-Value, this ID is encoded in the Key with the `m_` prefix. This constructs a Key, and stores the serialized meta information in Value.

Reviewer 2: ```suggestion
Each database or table is assigned a unique ID. As the unique identifier, when table data is encoded to Key-Value, this ID is encoded in the Key with the `m_` prefix. This constructs a key-value pair with the serialized metadata stored in it.
```

> Besides, TiDB also uses a dedicated (Key, Value) key pair to store the latest version number of the current all tables structure information. This key-value pair is global, and its version number is increased by 1 each time the state of the DDL operation changes. TiDB stores this key-value pair persistently in the PD Server with a key of "/tidb/ddl/global_schema_version", and Value is the version number value of int64 type. Refers to Google F1's  Online Schema change algorithm, TiDB keeps a background thread constantly checking whether the version number of the table structure information stored in the PD Server changes, and ensuring to get the changes of version in a certain time.

Reviewer 2: ```suggestion
In addition, TiDB also uses a dedicated (Key, Value) key-value pair to store the latest version number of all table structure information. This key-value pair is global, and `its version number is increased by 1 each time the state of the DDL operation changes. TiDB stores this key-value pair persistently in the PD server with the key of `/tidb/ddl/global_schema_version`, and Value is the version number value of the `int64` type. Inspired by Google F1's  Online Schema change algorithm, TiDB keeps a background thread that constantly checks whether the version number of the table structure information stored in the PD server changes. This thread also ensures that the changes of version can be obtained within a certain period of time.
```

> In addition, TiDB also uses a dedicated (Key, Value) key-value pair to store the latest version number of all table structure information. This key-value pair is global, and `its version number is increased by 1 each time the state of the DDL operation changes. TiDB stores this key-value pair persistently in the PD server with the key of `/tidb/ddl/global_schema_version`, and Value is the version number value of the `int64` type. Inspired by Google F1's  Online Schema change algorithm, TiDB keeps a background thread that constantly checks whether the version number of the table structure information stored in the PD server changes. This thread also ensures that the changes of version can be obtained within a certain period of time.

Reviewer 2: ```suggestion
In addition, TiDB also uses a dedicated (Key, Value) key-value pair to store the latest version number of all table structure information. This key-value pair is global, and its version number is increased by 1 each time the state of the DDL operation changes. TiDB stores this key-value pair persistently in the PD server with the key of `/tidb/ddl/global_schema_version`, and Value is the version number value of the `int64` type. Inspired by Google F1's  Online Schema change algorithm, TiDB keeps a background thread that constantly checks whether the version number of the table structure information stored in the PD server changes. This thread also ensures that the changes of version can be obtained within a certain period of time.
```

## Introduction to the SQL layer

TiDB's SQL layer, TiDB Server, is responsible for translating the SQL into Key-Value operation to the common distributed Key-Value storage layer TiKV, assembling the results returned by TiKV, and returning the query results to the client ultimately.

The nodes at this layer are stateless, the nodes themselves do not store data, and the nodes are completely equal.

> ### SQL algorithm

Reviewer 2: ```suggestion
### SQL computing
```

> The simplest solution is through the [mapping of table data to Key-Value](#mapping-of-table-data-to-key-value) as described in the previous section scheme, mapping SQL queries to KV queries, and then acquires the corresponding data through the KV interface and performs various computations.

Reviewer 2: ```suggestion
The simplest solution to SQL computing is the [mapping of table data to Key-Value](#mapping-of-table-data-to-key-value) as described in the previous section, which maps SQL queries to KV queries, acquires the corresponding data through the KV interface, and performs various computations.
```

> For example, `select count(*) from user where name = "TiDB"` such a SQL statement. It needs to read all the data in the table, then check if the `name` field is `TiDB`, and if so, returns this line. The process is as follows:

Reviewer 2: ```suggestion
For example, to execute the `select count(*) from user where name = "TiDB"` SQL statement, TiDB needs to read all data in the table, then checks whether the `name` field is `TiDB`, and if so, returns this row. The process is as follows:
```

> 1. construct the Key Range: all `RowID` in a table are in `[0,  MaxInt64)` range. According to the row data `Key` encoding rule, using `0` and `MaxInt64` can construct a `[StartKey, EndKey)` range that is left-included, right-excluded.

Reviewer 2: ```suggestion
1. Construct the Key Range: all `RowID` in a table are in `[0,  MaxInt64)` range. According to the row data `Key` encoding rule, using `0` and `MaxInt64` can construct a `[StartKey, EndKey)` range that is left-inclusive and right-exclusive.
```

> 2. scan Key Range: read the data in TiKV according to the key range constructed above.

Reviewer 2: ```suggestion
2. Scan Key Range: read the data in TiKV according to the key range constructed above.
```

> 3. filter data: for each row of data read, calculate `name = "TiDB"` expression. Returns up this line if true, otherwise discards this line of data.

Reviewer 2: ```suggestion
3. Filter data: for each row of data read, calculate the `name = "TiDB"` expression. If the result is true, return to this row. If not, skip this row.
```

> 4. calculate `Count(*)`: for each line that meets the requirements, accumulate to the result of `Count(*)`.

Reviewer 2: ```suggestion
4. Calculate `Count(*)`: for each row that meets the requirements, add up to the result of `Count(*)`.
```

**The entire process is illustrated as follows:**

![naive sql flow](/media/tidb-computing-native-sql-flow.jpeg)

This solution is intuitive and feasible, but has some obvious problems in a distributed database scenario.

> - As the data is being scanned, each row is read out of TiKV via a KV operation at least once RPC overhead, which can be very high if there is a lot of data to scan.

Reviewer 2: ```suggestion
- As the data is being scanned, each row is read out of TiKV via a KV operation with at least one RPC overhead, which can be very high if there is a large amount of data to be scanned.
```

> - Not all rows meet the filter criteria `name = "TiDB"`. If the conditions are not met, they are unnessary to be read out.

Reviewer 2: ```suggestion
- It is not applicable to all rows. Data that doesn't meet the conditions doesn't need to be read.
```

> - The value of the rows that meet the requirements doesn't mean anything, in fact, all needed here is the information of how many rows of data.

Reviewer 2: ```suggestion
- The value of the rows that meet the conditions is meaningless. In fact, all needed here is just the number of rows.
```

### Distributed SQL operations

> To solve the above problem, the computation should need to be as close to the storage node as possible to avoid a large number of RPC calls. First of all, the SQL predicate condition `name = "TiDB"` should be pushed down to the storage node for computation, so that only valid rows are returned, avoiding meaningless network transfers. Then, the aggregation function `Count(*)` can also be pushed down to the storage nodes for pre-aggregation, and each node only has to return a result of `Count(*)`, and the SQL layer will sum up the `Count(*)` result returned by each node.

Reviewer 2: ```suggestion
To solve the problems above, the computation should be as close to the storage node as possible to avoid a large number of RPC callings. First of all, the SQL predicate condition `name = "TiDB"` should be pushed down to the storage node for computation, so that only valid rows are returned, which avoids meaningless network transfers. Then, the aggregation function `Count(*)` can also be pushed down to the storage nodes for pre-aggregation, and each node only has to return a result of `Count(*)`. The SQL layer will sum up the `Count(*)` results returned by each node.
```

> The following is a schematic representation of the data returned layer by layer.

Reviewer 2: ```suggestion
The following image shows how data returns layer by layer:
```

![dist sql flow](/media/tidb-computing-dist-sql-flow.png)

> ### SQL layer architecture

Reviewer 2: ```suggestion
### Architecture of the SQL layer
```

> With the above example, I hope you have a basic understanding of how SQL statements are handled. In fact, TiDB's SQL layer is much more complex, with many modules and layers. The following diagram lists the important modules and calling relationships:

Reviewer 2: ```suggestion
The previous sections introduce some functions of the SQL layer and I hope you have a basic understanding of how SQL statements are handled. In fact, TiDB's SQL layer is much more complicated, with many modules and layers. The following diagram lists the important modules and calling relationships:
```

![tidb sql layer](/media/tidb-computing-tidb-sql-layer.png)

> The user's SQL request is sent to TiDB Server either directly or via `Load Balancer`. TiDB Server will parse `MySQL Protocol Packet`, getting the content of requests, parsing syntax and semantic analysis of SQL, developing and optimizing query plans, executing a query plan , fetching and processing the data. The data is all stored in the TiKV cluster, so in this process, TiDB Server needs to interact with the TiKV and gets the data. Finally, TiDB Server needs to return the query results to the user.

Reviewer 2: ```suggestion
The user's SQL request is sent to TiDB Server either directly or via `Load Balancer`. TiDB Server will parse `MySQL Protocol Packet`, get the content of requests, parse the SQL request syntactically and semantically, develop and optimize query plans, execute a query plan, get and process the data. All data is stored in the TiKV cluster, so in this process, TiDB Server needs to interact with TiKV and gets the data. Finally, TiDB Server needs to return the query results to the user.
```

---
title: TiDB database computation
summary: Understand the computation layer of the TiDB database.
category: introduction

> aliases: ['/docs/dev/tidb-computing/']

Reviewer 2: ```suggestion
```
